feat: Add Unit Tests for Feature Auth

aslamhossin opened the pull request:
This pull request introduces changes to the codebase to implement unit tests for the AuthRepository class. The changes made include:

 - Added mapper extension function for AuthToken in Domain layer inside ResponseTokenDTO file.
 - Implemented the AuthRepository unit test using the mockk framework to mock dependencies and verify the behavior of the AuthRepository class.

 The introduction of unit tests for the AuthRepository class will improve the reliability and stability of the application

**Acceptance criteria:**  
- [ ] PR title includes Jira ticket reference
- [ ] Follows UI patterns from [UI manifesto](https://github.com/monstar-lab-oss/ui-manifesto)
- [ ] Loading/error/empty states taken into account
- [ ] Status of Jira ticket updated
- [ ] Removed commented out code  
- [ ] Removed unnecessary logging
- [ ] Tested flows in airplane mode and bad connection
- [ ] Updated documentation (please add a link if other than README 😊)

**Related to other PRs:**  

**Notes for reviewers:**

levinzonr commented on app/src/main/java/com/monstarlab/core/mapper/Mapper.kt:
```diff
@@ -0,0 +1,9 @@
+package com.monstarlab.core.mapper
+
+/**
+ * Created by aslam on 08,March,2023
+ */
+interface Mapper<I, O> {
+    fun map(input: I): O
+}
+
```
From my experience using the Mapper interface doesn't provide any value to it and it the end is a bit annoying.
In this template we ended up using extensions functions on the DTO to do the mapping, which is more straightforward IMO.

You can check [ResourceDTO](https://github.com/monstar-lab-oss/android-template/blob/main/app/src/main/java/com/monstarlab/features/resources/data/api/dtos/ResourceDTO.kt#L18) for the example

aslamhossin replied:
This approach using the interface and mapper implementation is more flexible and reusable. It enables you to define a generic mapping interface and multiple mapper implementations for various types of data. This approach is well-suited for larger projects that necessitate extensive data mapping and may have multiple data sources or layers.

In contrast, the approach you are proposing, which employs extension functions, is more succinct and user-friendly. It provides a rapid method for mapping data without creating supplementary classes or implementing an interface. This approach is suitable for smaller projects or components of larger projects where data mapping is not the primary focus.

Should I consider changing the implementation? What do you think?

levinzonr replied:
I would say in most scenarios having an extension function will be enough. 
Yes, you won't be able to inject the extension function, but that's about the only limitation I see, it doesn't remove any flexibility or reusability.

Extensions are as reusable as the Mapper interface. Having this much boilerplate to just enable Mapping in the repository does not appeal to me. So in the end yes, I think we should go with the extension function for the purpose of this template. 

As soon as the developer sees the need in having such a structure and the Mapping is very complex it deserves its own component, then yes extension function doesn't make sense and  this of course is the way to go

levinzonr commented on app/src/main/java/com/monstarlab/features/auth/domain/AuthRepository.kt:
```diff
@@ -1,15 +1,16 @@
 
 class AuthRepository @Inject constructor(
     private val api: AuthApi,
+    private val authTokenMapper: AuthTokenMapper
```
With the Mappers discussion aside, this injection doesn't bring any benefits since you inject an implementation and not an Interface. And since its a Singleton it can just be an `object`

aslamhossin replied:
Great observation! The purpose of this mapper class is to provide mapping functionality between two different object types, making the code more modular, testable, and maintainable. This approach is not primarily intended for dependency injection as the mapper interface is generic.

However, creating a singleton using an `object` keyword has a downside - it is not easy to mock the `AuthTokenMapper` during unit testing. If you need to test classes that depend on the `AuthTokenMapper`, using dependency injection to provide a mock implementation may be a better approach.

levinzonr replied:
Even with the mappers as is, I don't see a reason why mock such a simple thing if you can use its Implementation directly. It doesn't have any dependencies on its own + there is no test for the Mapper Implementation. So if someone changes the Mapper implementation and makes a mistake - The test will pass anyways

levinzonr commented on app/src/test/java/com/monstarlab/features/auth/domain/AuthRepositoryTest.kt:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    @Test
+    fun testLoginReturnsExpectedAuthToken() = runBlocking {
+        // Arrange
+        val email = "aslam@test.com"
+        val password = "password"
+        val expectedTokenResponseDTO = TokenResponseDTO("token")
+        val expectedAuthToken = AuthToken("token")
+        coEvery { mockAuthApi.postLogin(email, password) } returns expectedTokenResponseDTO.also {
+            every {
+                mockAuthTokenMapper.map(it)
+            } returns expectedAuthToken
+        }
+
+        // Act
+        val actualAuthToken = authRepository.login(email, password)
+
+        // Assert
+        assertEquals(expectedAuthToken, actualAuthToken)
+        coVerify { mockAuthApi.postLogin(email, password) }.also {
+            verify {
+                mockAuthTokenMapper.map(expectedTokenResponseDTO)
+            }
+        }
+    }
```
If you think about this test doesn't really test the real implementation
The repo has two functions:
1. call the API
2. Map the DTO to the domain class

Both those things are mocked so the only thing under the test here is that these mocks are actually called
The mapping mock can be removed for sure, there is no reason to mock it

aslamhossin replied:
Regarding this test, I have mocked the API call and map function, and then performed an action on the unit that needs to be tested. After that, assertions and verification were used to check that the mock endpoint was called correctly. For better understanding, I have pushed a change. Please take a look.

levinzonr commented on app/src/test/java/com/monstarlab/features/auth/domain/AuthRepositoryTest.kt:
```diff
@@ -0,0 +1,79 @@
+        }
+    }
+
+    /**
+     * A more appropriate approach would be to modify the repository layer to handle exceptions
+    that may occur during login operations, and then write tests that verify that the repository
+    layer properly handles these exceptions. This way, the code is more resilient and can handle
+    unexpected situations that may occur during the login process.
+     * Test to ensure that a network error is thrown when logging in and
+    the repository layer does not throw one.
+     */
+    @Test
+    fun testLoginNetworkErrorThrowsException() = runBlocking {
+        // The repository should throw a Network Error when there is a problem with the network connection.
+    }
+
+    // Test to ensure that a server error is thrown when logging in and the repository layer does not throw one.
+    @Test
+    fun testLoginServerErrorThrowsException() = runBlocking {
+        // The repository should throw a Server Error when there is a problem with the server.
+    }
+
+}
```
The repository technically  layer technically throws the exceptions :)
They will be thrown [here first](https://github.com/aslamhossin/android-template/blob/feature/auth_unit_test/app/src/main/java/com/monstarlab/core/network/errorhandling/ApiErrorInterceptor.kt) 
and since there is no try-catch here, the repository will throw it as well, then it will be either handled in the case or in a ViewModel

So tests like these will  actually be useful only if you are using a proper MockWebServer and enqueue real Responses with the `ApiErrorInterceptor` setup

levinzonr commented on app/src/test/java/com/monstarlab/features/auth/domain/AuthRepositoryTest.kt:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    @Test
+    fun testLoginReturnsExpectedAuthToken() = runBlocking {
```
Please use `runTest` instead of `runBlocking`.  Its available in the coroutines-test artifact

levinzonr reviewed (changes requested):
Thank you for your contribution! :]
I left a few comments, feel free to reach me on slack or tag here in case you have any questions :)

aslamhossin commented:
> Thank you for your contribution! :] I left a few comments, feel free to reach me on slack or tag here in case you have any questions :)

@levinzonr Thank you for taking the time to review my pull request. I have addressed some of the feedback and made changes based on your suggestions. Could you please review the pull request again?

levinzonr commented on app/build.gradle.kts:
```diff
@@ -111,6 +111,10 @@ dependencies {
     testImplementation(Libs.Test.junit)
+    testImplementation(Libs.Test.mockkAndroid)
+    testImplementation(Libs.Test.mockkAgent)
+    testImplementation ("org.jetbrains.kotlinx:kotlinx-coroutines-test:1.6.4")
```
Please put the dependency in the Libs.Test.. thingie

aslamhossin replied:
@levinzonr I've made the necessary change. Please review it again.
